### DISCUSSION TITLE
Correct way to deserialize an external JSON

### DISCUSSION THREAD

**Discussion by AndryHTC** [serverpod/serverpod#515] [Q&A] [answered]
Hi!
I'm still having trouble understanding how to properly deserialize an external JSON file.

I noticed that there is a `ProtocolClass.fromJson(jsonSerialization, serializationManager)` method in the generated protocol class, but I don't know what to pass as second parameter.

I can also deserialize a JSON using `Protocol().deserialize(json, GeneratedProtocol)`, but I'm not sure if this is the preferred method because it imports all the generated protocols.

And... Is there a way to deserialize a list of JSONs? In the past, I used `Protocol().deserialize(jsonList, List<GeneratedProtocol>)` for this purpose.

The documentation does not help a lot with it, I hope we can clearly explain how we can deserialize JSON into protocol classes as designed by the DEV.

**Comment by vlidholt** [comment, marked as answer]
You can retrieve the `SerializationManager` from your `Serverpod` object. E.g. `Serverpod.instance!.serializationManager`. The `SerializationManager` has knowledge about all the classes from the yaml-files (plus any extra customly serialized classes you may have registered).

## Deserialization / decoding
If you know the type of the object you want to deserialize, use the `SerializationManagers` `deserialize` method. E.g:
```dart
// From JSON structure in memory (i.e. lists, maps, primitives).
var myObject = Serverpod.instance!.serializationManager.deserialize<MyType>(json);

// From JSON encoded string
var myObject = Serverpod.instance!.serializationManager.decode<MyType>(jsonString);
```

## Serialization / encoding
Typically, just call the `toJson` method to serialize an object. However, that will not encode DateTime, Duration, etc. Those will be saved as their original types in the JSON structure.

To encode these to a `String`, use the `SerializationManager`'s `encode` method. E.g.:
```dart
var jsonString = Serverpod.instance!.serializationManager.encode(myObject);
```

## If the type is unknown
If the type will be unknown when decoding the object (this is true for instance in streams), then use the `SerializationManager`'s `encodeWithType` and `decodeWithType` methods. It will wrap the encoded JSON with the class name, which can be used to correctly decode the serialized object.

**Comment by jason-downs** [comment]
Stuff like this really needs to be in the official Serverpod documentation. There's too much basic functionality that is missing from those docs right now despite already having implementations, such as enum serialization in yaml files, to take another example. Hopefully there is a plan soon to move some of these Q&As and other undocumented features into the docs.
